src/server.c

```diff
@@ -3445,15 +3451,17 @@ void call(client *c, int flags) {
 
     /* populate the per-command statistics that we show in INFO commandstats. */
     if (flags & CMD_CALL_STATS) {
-        real_cmd->microseconds += duration;
-        /* In case we are in the process of unblocking a client on keys,
-         * we are going to reprocess a command which was already process once so we do not want to account
-         * for the command twice */
-        if (!(c->flags & CLIENT_UNBLOCKED))
+        /* We want to increment the number of calls stats only once.
+         * why? because the blocked command can also be timedout or cancel, and we would still
+         * want to count for it. however we do want to accumulate the command processing time
+         * only after it is unblocked. */
+        if (!(c->flags & CLIENT_REPROCESSING_COMMAND))
```

Comment:
CLIENT_UNBLOCKED will also be set for posponed client which will not reprocess the command. Personally I so not understand why we even want to report commands which did not complete (eg timedout) we could maybe add a function to accumulate them to the statistics when a client is timedout (like the "updateStatsOnUnblock" we can have an updateStatsOnTimeout. But I wanted to keep the current behavior similar to what it is today.

Reply:
postponed clients **do** re-process the command, it'll get to `processPendingCommandAndInputBuffer` and call `processCommand` again for the postponed command.

maybe for perspective, think of module blocked clients.. they can do some work on the original execution, then do some incremental work from time to time, and then another bulk of code when being unblocked.
the same could be argued to apply for stream blocked commands and others.

i do think we wanna increment the fact the command was called (as soon as it got to `call()`), and measure all the fragments of time it took.

Reply:
@oranagra correct me if I am wrong. postpone client would start processing the command but will quit one it gets into the check of: blockPostponeClient - which is before it got into the call(). so it would not have been recorded anywhere.
If I used the CLIENT_UNBLOCKED  here instead of CLIENT_REPROCESSING_COMMAND   the postpone client would NOT account once it will reprocess the command, which is why I added this flag (otherwise I would not need it )
IMO I would prefer what I suggested earlier which is to not account anything (including maybe even the monitor) for blocked commands and only account them once the command is executed.

Reply:
ohh, ok. postponed commands are "reprocessed" from the perspective of `processCommand()`, but not from the perspective of `call()`. so you're right, we must not skip them.
i think that when i gave that comment, i looked at a diff from a previous commit (of this PR) and i wondered why you changed the flag, but now i see that entire check is new (added in this PR).

i think i do prefer to keep counting commands that got to `call` even if eventually timed out, but i suppose i'm open to compromise on that if it conflicts with many things and makes things much simpler.

Reply:
@oranagra we had some offline discussions about that.
As much as I tried, merging the logic of blocked commands and postpone clients is harder than I thought.
I think the main difference is that postpone clients are not really calling the command, but rather being postpone "on the way" to call the command as apposed to other blocked commands which decide to be blocked somewhere during the command execution. I do not think this is the last word on the subject and I plan to do farther refactor (modules etc...)
Maybe I can document our thoughts on the top commit and we leave it to the next steps in order to push this long going PR?

Reply:
@oranagra, @madolson. After your comment about the monitor command, I **DID** apply the following logic:
When a client is blocked during processing of some command it will **NOT** update any statistic and will not report to the monitor. that will help keep all blocking flows  with about the same logic.
This also implies that when we have many clients blocked on blocking command we will have no way to get that from the command stats (user can still get the information from the client list for example). 
I want to have this first approved by you and then I will update the top comment.

Reply:
I don't think command stats should have the information about blocking. We should likely have a separate field indicating the number of blocked clients along with client list which indicates clients are blocked.

Reply:
@madolson I think my last commits does exactly that. we already have info fields for the number of blocked clients and the number of blocked keys. We also have an indication for blocked clients on the client list.

Reply:
> This also implies that when we have many clients blocked on blocking command we will have no way to get that from the command stats (user can still get the information from the client list for example).
I want to have this first approved by you and then I will update the top comment.

approved, please mention it in the top comment and resolve this discussion.

Reply:
done.